### PR TITLE
feat(kic 2.0) support selector tags

### LINKED ISSUE #1292: Support selector tags in KIC 2.0
The line below always implies that selector tags are configured as empty.

https://github.com/Kong/kubernetes-ingress-controller/blob/347ed5fb48a6f3e74b537e56e82259bdb0d82b1a/railgun/internal/proxy/clientgo_cached_proxy_resolver.go#L216

### PR DESCRIPTION
**Pull request by tharun208**
Signed-off-by: Tharun <rajendrantharun@live.com>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1292

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: railgun/manager/run.go [hunk 1]
```diff
@@ -91,12 +91,18 @@ func Run(ctx context.Context, c *config.Config) error {
...
+
 	// configure the kong client
 	kongConfig := sendconfig.Kong{
 		URL:               c.KongAdminURL,
 		FilterTags:        c.FilterTags,
 		Concurrency:       c.Concurrency,
 		Client:            kongClient,
+		HasTagSupport:     isTagExists,
 		PluginSchemaStore: util.NewPluginSchemaStore(kongClient),
 	}
```
Comment (mflendrich) [inline railgun/manager/run.go:107]: This PR implements tags support in a way very similar to how KIC 1.x did, but let's use this opportunity to simplify the implementation. We could:

1. mark the `sendconfig.Kong.HasTagSupport` field as deprecated in KIC 2.x
2. make `sendconfig` expect that, if Kong doesn't support tags, then the `FilterTags` list be empty

If this sounds right, let's:

1. rename the `HasTagSupport` field to `DeprecatedHasTagSupport` and add a documentation comment like `// DeprecatedHasTagSupport is not used in KIC 2.x. If the gateway instance does not support tags, pass an empty FilterTags slice instead.`
2. rewrite the `kongClient.Tags.Exists(ctx)` check above to something like:
    ```go
    var filterTags []string
    if ok, err := kongClient.Tags.Exists(ctx); err != nil {
      setupLog.Error(err, "cannot able to get tags from Kong Admin API client")
      return err
    } else if ok {
      setupLog.Info("tag filtering enabled", "tags", c.filterTags)
      filterTags = c.FilterTags
    }
    ```
3. drop the `HasTagSupport` field from the `kongConfig` initialization in line 100 and use `filterTags` from the code snippet above
4. update the flag description by appending `This setting will be silently ignored if the Kong instance has no tags support.`

File: pkg/sendconfig/sendconfig_test.go [hunk 2]
```diff
@@ -128,3 +128,51 @@ func Test_updateReportingUtilities(t *testing.T) {
 	assert.True(t, hasSHAUpdateAlreadyBeenReported([]byte("yet-another-fake-sha")))
 	assert.True(t, hasSHAUpdateAlreadyBeenReported([]byte("yet-another-fake-sha")))
 }
+
```
Comment (mflendrich) [inline pkg/sendconfig/sendconfig_test.go:131]: Please add an integration test that will:
- specify a custom list of tags
- verify that entries created in Kong have the tags appropriately set
- verify that preexisting entries not matching the filter tag are not affected

File: CHANGELOG.md [hunk 3]
```diff
@@ -37,6 +37,8 @@ released and the release notes may change significantly before then.
 
 #### Added
 
+- support for selector tags in KIC 2.0.
```
Comment (mflendrich) [inline CHANGELOG.md:40]:
```suggestion
- adds support for selector tags (filter tags) refactored wrt. KIC 1.x
```

File: railgun/manager/run.go (hunk 1 above)
Comment (tharun208) [reply railgun/manager/run.go]: addressed

File: pkg/sendconfig/sendconfig_test.go (hunk 2 above)
Comment (tharun208) [reply pkg/sendconfig/sendconfig_test.go:131]: @mflendrich, can you explain more about the 3rd point?

Comment (mflendrich) [reply pkg/sendconfig/sendconfig_test.go:131]: In DB mode, when KIC is run with a non-empty `--filter-tags`, it causes the following:
- all resources that KIC creates in Kong Admin API have the tag specified
- for sync purposes KIC sees only those resources in the Kong Admin API that have the tag specified

Therefore, resources not matching the tag filter should be "invisible" to KIC and not affected by KIC.

I'm asking for a test that confirms the above behavior. For example:
- KIC and Kong proxy start up
- the test manually creates a route X in Admin API with a "managed-by-ingress-controller" tag
- the test manually creates a route Y in Admin API with a "managed-by-something-else" tag
- KIC tries to sync
- the test verifies that, after sync, X no longer exists in the cluster but Y remains

makes sense?

Comment (mflendrich) [reply pkg/sendconfig/sendconfig_test.go:131]: @hbagdi anything to add/comment wrt. tag support behavior?

Comment (tharun208) [reply pkg/sendconfig/sendconfig_test.go:131]: @mflendrich sure will come with the test 👍🏻

File: internal/ingress/controller/kong.go [hunk 4]
```diff
@@ -100,13 +103,3 @@ func (n *KongController) fetchCustomEntities() ([]byte, error) {
...
-
-// getIngressControllerTags returns a tag to use if the current
-// Kong entity supports tagging.
-func (n *KongController) getIngressControllerTags() []string {
-	var res []string
-	if n.cfg.Kong.HasTagSupport {
-		res = append(res, n.cfg.Kong.FilterTags...)
-	}
-	return res
-}
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:112]: `getIngressControllerTags` is only used by KIC 1.x (and is a part of a mechanism deprecated in 2.x), so I think there's no value in moving it to `package sendconfig`, however there's a cost - `package sendconfig` being burdened by a 1.x-specific enabling mechanism that will be phased out very soon.

Therefore, please move this function implementation back to this file.

Rule of thumb: "don't scatter deprecated concepts around the codebase, keep them as isolated as possible"

File: railgun/manager/run.go (hunk 1 above)
Comment (mflendrich) [reply railgun/manager/run.go]: My previous code suggestion in point 2 was buggy :roll_eyes: sorry about that.

I've updated the suggestion and committed the change directly to your branch.

File: internal/ingress/controller/kong.go (hunk 4 above)
Comment (tharun208) [reply internal/ingress/controller/kong.go:112]: sure! got it 👍🏻

Comment (mflendrich) [reply internal/ingress/controller/kong.go]: Oh. I realized that would save us a round trip if I did it right away, and I've pushed a commit performing this already :sweat_smile:

**Comment by mflendrich** [comment]
@tharun208 seems that we've started changing this PR at the same instant. Let me finish the modifications and merge recent changes into this PR, sounds good? There's no other action required on your side on this PR, the change looks good :+1:

**Comment by tharun208** [comment]
> @tharun208 seems that we've started changing this PR at the same instant. Let me finish the modifications and merge recent changes into this PR, sounds good? There's no other action required on your side on this PR, the change looks good 👍

I was about to address the feedback changes. thanks for doing it 🙏

**Review by mflendrich — APPROVED** [review]

**Pull request merged**